modules/markdown/assets/js/components/markdown-live-preview.jsx

```diff
@@ -0,0 +1,193 @@
+import escape from 'lodash/escape';
+import MarkdownIt from 'markdown-it';
+
+/**
+ * Internal dependencies
+ */
+import { saveCaretPosition } from '../utils/caret-management';
+
+const {
+	createElement
+} = window.wp.element;
+
+const markdownIt = new MarkdownIt( 'zero' ).enable( [
+	'heading',
+	'emphasis',
+	'backticks',
+] );
+
+// Redefines the rules applied by the parser to render each token.  This adds
+// to each token's content, the delimiter the user used (which the parser
+// obviously removes in the resulting HTML)
+const setupMarkdownParser = function() {
+	// Adds `_` or `*` to the beginning of the em tag
+	markdownIt.renderer.rules.em_open = function( tokens, idx ) {
+		const token = tokens[ idx ];
+		return `<em>${ token.markup }`;
+	};
+	// Adds `_` or `*` to the end of the em tag
+	markdownIt.renderer.rules.em_close = function( tokens, idx ) {
+		const token = tokens[ idx ];
+		return `${ token.markup }</em>`;
+	};
+	// Adds `__` or `**` to the beginning of the strong tag
+	markdownIt.renderer.rules.strong_open = function( tokens, idx ) {
+		const token = tokens[ idx ];
+		return `<strong>${ token.markup }`;
+	};
+	// Adds `__` or `**` to the end of the strong tag
+	markdownIt.renderer.rules.strong_close = function( tokens, idx ) {
+		const token = tokens[ idx ];
+		return `${ token.markup }</strong>`;
+	};
+	// Wraps inline code tokens with ```
+	markdownIt.renderer.rules.code_inline = function( tokens, idx ) {
+		const token = tokens[ idx ];
+		return `<code>${ token.markup }${ escape( token.content ) }${ token.markup }</code>`;
+	};
+	// Adds `#`s to the beginning of the heading content
+	markdownIt.renderer.rules.heading_open = function( tokens, idx ) {
+		const token = tokens[ idx ];
+		const inline_token = tokens[ idx + 1 ];
+		const text_token = inline_token.children[ 0 ];
+		if ( text_token ) {
+			text_token.content = ` ${ text_token.content }`;
+		}
+		return `<${ token.tag }>${ token.markup }`;
+	};
+};
+
+const renderHTML = function( source ) {
+	if ( source ) {
+		return markdownIt.render( source )
+		// The MarkdownIt renderer adds new lines after each paragraph/heading
+		// tag. This affects the restoration of the caret position, so we need
+		// to remove them
+			.split( /\n/ ).join( '' )
+			.trim();
+	}
+};
+
+const stripTrailingNewLines = function( text ) {
+	if ( text ) {
+		return text.replace( /\n{1,2}$/, '' );
+	}
+};
+
+const triggerOnChange = function( evt, source ) {
+	if ( this.props.onChange ) {
+		// Clone event with Object.assign to avoid
+		// "Cannot assign to read only property 'target' of object"
+		evt = Object.assign( {}, evt, {
+			target: {
+				value: source
+			}
+		} );
+		this.props.onChange( evt );
+	}
+};
+
+const sourceIsEmpty = function( source ) {
+	return ! source || '' === source.trim();
+};
+
+const ignoreLastInput = function( source ) {
+	return endsWith( source, ' ' ) || endsWith( source, ' ' );
+};
+
+const emptyState = '<p></p>';
+
+const emitChange = function( evt ) {
```

Comment:
What was the rationale for keeping this function outside the component class?

Reply:
Independently of that, why isn't `emitChange` bound in the class constructor?

Reply:
Yet another question :) : why the name `emitChange`? To me it reads as if the function is doing more than it is, especially compared with some similar handlers found in core Gutenberg.

Reply:
Answering your first two questions: I guess this was just personal preference. I tend to not expose those functions that aren't supposed to be consumed out of the class. I don't know what is the convention you'd use here, but this was my reasoning.
And regarding the name... Yeah, probably that wasn't my best choice :sweat_smile: I'll give it another shot.